### ISSUE TITLE
Fields API should enforce unique names

### ISSUE THREAD

**Issue by kjohnson** [impress-org/givewp#5594]
Each node in the AST should have a unique identifier, which should be enforced when a new node is added to the tree.

**Comment by kjohnson** [comment]
The individual nodes need a reference to the `root` node in order to trigger at the top of the tree - otherwise, each collection can only check its own children.

Perhaps adding a `getRoot()` method which would return a set `root` reference or itself if one is not set. This would give each child reference to the `root` node.

**Comment by kjohnson** [comment]
After speaking with @JasonTheAdams, we have decided to simplify collision detection by assuming the `root` collection as the entry point for new nodes, as opposed to inserting nodes directly into a nested collection. This avoids needing to weave a reference to the root collection down each branch of the tree.

We discussed the scenario where a developer uses a collection to group fields (ie checkbox group). The collection can be created and nodes can be inserted on that collection, as at that time it is a a root collection. Once the collection is inserted as a node in another tree, the receiving tree is the shared root collection.

Additionally, inserted nodes should be be checked for collision detection prior to accepting an incoming node. This is important to prevent an inserted node from colliding with itself. For collections, each node should be checked for collision in the accepting tree (deep collision detection).
